### PR TITLE
fix(auth): Refresh id token when signIn happens

### PR DESCRIPTION
**Pull request by royjit**
*Issue #2922 

*Description of changes:*

Identity id is not refreshed if we immediately try to fetch after signin. This change refreshes the id token before returning from signin callback by calling aws credentials.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1186,6 +1186,7 @@ workflows:
               only:
                 - develop
                 - main
+                - royjit/refresh_id_token
```
Comment (palpatim) [inline .circleci/config.yml:1189]:
```suggestion
```

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift [hunk 2]
```diff
@@ -523,13 +523,18 @@ final public class AWSMobileClient: _AWSMobileClient {
...
                     self.performHostedUISuccessfulSignInTasks(disableFederation: hostedUIOptions.disableFederation, session: session, federationToken: federationToken!, federationProviderIdentifier: federationProviderIdentifier, signInInfo: &signInInfo)
-                    self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: signInInfo)
-                    completionHandler(.signedIn, nil)
                     if self.pendingGetTokensCompletion != nil {
                         self.tokenFetchLock.leave()
                     }
                     self.pendingGetTokensCompletion?(self.getTokensForCognitoAuthSession(session: session), nil)
                     self.pendingGetTokensCompletion = nil
+
+                    // Invoke credentials inorder to refresh the id token before returning signedin
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift:532]:
```suggestion
                    // Invoke `credentials` in order to refresh the id token before returning signedIn
```

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift [hunk 3]
```diff
@@ -543,10 +548,16 @@ final public class AWSMobileClient: _AWSMobileClient {
                             if let session = task.result {
                                 self.performUserPoolSuccessfulSignInTasks(session: session)
                                 let tokenString = session.idToken!.tokenString
-                                self.mobileClientStatusChanged(userState: .signedIn,
-                                                               additionalInfo: [self.ProviderKey:self.userPoolClient!.identityProviderName,
-                                                                                self.TokenKey:tokenString])
-                                completionHandler(.signedIn, nil)
+
+                                // Invoke credentials inorder to refresh the id token before returning signedin
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift:552]:
```suggestion
                                // Invoke `credentials` in order to refresh the id token before returning signedIn
```

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 4]
```diff
@@ -235,10 +235,15 @@ extension AWSMobileClient {
                     self.federationProvider = .userPools
                     self.performUserPoolSuccessfulSignInTasks(session: result)
                     let tokenString = result.idToken!.tokenString
-                    self.mobileClientStatusChanged(userState: .signedIn,
-                                                   additionalInfo: [self.ProviderKey:self.userPoolClient!.identityProviderName,
-                                                                    self.TokenKey:tokenString])
-                    self.invokeSignInCallback(signResult: SignInResult(signInState: .signedIn), error: nil)
+                    
+                    // Invoke credentials inorder to refresh the id token before returning signedin
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift:239]:
```suggestion
                    // Invoke `credentials` in order to refresh the id token before returning signedIn
```

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 5]
```diff
@@ -263,9 +268,14 @@ extension AWSMobileClient {
             // At the end of operation if there is an error anywhere in the flow, we return it back to the developer; else return a successful signedIn state.
             defer {
                 if error == nil {
-                    self.mobileClientStatusChanged(userState: .signedIn,
-                                                   additionalInfo: [self.ProviderKey:providerName, self.TokenKey: token])
-                    completionHandler(UserState.signedIn, nil)
+                    // Invoke credentials inorder to refresh the id token before returning signedin
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift:271]:
```suggestion
                    // Invoke `credentials` in order to refresh the id token before returning signedIn
```

**Review by palpatim — COMMENTED** [review]
Some changes. Will connect offline to review the codepath and ensure this won't result in a deadlock

**Comment by royjit** [comment]
This PR caused delay for the signIn callback to return if there are many getToken api calls are pending. Created another PR - https://github.com/aws-amplify/aws-sdk-ios/pull/3236

**Pull request closed without merging**